### ISSUE TITLE
performance Issue in reading data from monodb using MongodbConfigBuilder 

### ISSUE THREAD

**Issue by Velmuruganms12** [Stratio/Spark-MongoDB#106]
We are trying to use MongodbConfigBuilder for reading mongodb data to mongoRDD as shown below. But when our table has million of records when loading mutiple tables its delays whole performance. Reading each time takes lot of time.

Can we apply select or filter condition to load only few rows of mongo table or apply filter to load few rows to improve performance.

      val mcInputBuilder = MongodbConfigBuilder(Map(MongodbConfig.Host -> List(MongoHost + ":" + MongoPort), MongodbConfig.Database -> Database, MongodbConfig.Collection -> tableName,
        MongodbConfig.SamplingRatio -> 1.0, MongodbConfig.WriteConcern -> MongodbWriteConcern.Normal, MongodbConfig.SplitSize -> 1000, MongodbConfig.ConnectionsPerHost -> "100"))
      val mongoRDD = sqlContext.fromMongoDB(mcInputBuilder.build())
      mongoRDD.registerTempTable(tableName)

**Comment by AdrianP-** [comment]
I have the same problem. In fact, every query has an incredible slow performance.

**Comment by darroyocazorla** [comment]
Hi @Velmuruganms12,

First of all, you can take a look at available [ SQL and DataFrame API examples](https://github.com/Stratio/Spark-MongoDB/tree/master/spark-mongodb-examples/src/main/scala/com/stratio/datasource/mongodb/examples); what you are getting when you execute sqlContext.fromMongoDB(mcInputBuilder.build()) is a [DataFrame](https://spark.apache.org/docs/1.5.1/api/java/org/apache/spark/sql/DataFrame.html) You can apply transformations to it without effectively loading nor transforming the data, only after applying a Spark action ( such as collect, show, etc...) will operations be performed.  That means you could apply a filter transformation thus avoiding the load of most of your table data.

Also, increasing the MongoDB cursor batch size is a good practice, specially helpful when you are forced to process the whole table; it can be setted by adding the spark-mongodb cursorBatchSize option to your config. This parameter has been included since the version 0.11.0. I guess you are not using that version checking out your conf, which include MongodbWriteConcern.Normal (not supported in 0.11.0 and not required to read data), so you can try upgrading spark-mongodb.to 0.11.0.

If it won't fix the issue, could you please provide more info to reproduce the issue? Which version of Spark and spark-mongodb are you using? Standalone/ Mesos/ Yarn? Are you using a single replica set? What kind of operations are you doing? How many tasks are generated?... All these things would be useful to try to provide specific SparkSQL/spark-mongodb config in order to help to increase the performance.

I hope it might be helpful

**Comment by Velmuruganms12** [comment]
Our issue is not with using Dataframe  or transforming its data, our issue is loading million record in to dataframe from mongo which takes lot of time. Most of our table has million records in it, if we are going to use that using dataframe whole table needs to be loaded into in memory.

We tried increasing batch size by changing the MongodbConfig.SplitSize -> 100000 but performance did not improve.

We use Cloudera 5.5.1 Yarn Spark 1.5 and Mongo 3.2. We use single instance of mongo DB, Cloudera spark cluster with 5  slaves 16GB aws machine. For loading 10 million record table into memory it took around 8 min 40 sec.

**Comment by darroyocazorla** [comment]
As I said before, 

> "Increasing the MongoDB cursor batch size is a good practice, specially helpful when you are forced to process the whole table; it can be setted by adding the spark-mongodb cursorBatchSize option to your config. This parameter has been included since the version 0.11.0"

So, you can replace your config with:

` val mcInputBuilder = MongodbConfigBuilder(Map(MongodbConfig.Host -> List(MongoHost + ":" + MongoPort), MongodbConfig.Database -> Database, MongodbConfig.Collection -> tableName,
    MongodbConfig.SamplingRatio -> 0.4, MongodbConfig.CursorBatchSize -> 1000, MongodbConfig.SplitSize -> 100, MongodbConfig.ConnectionsPerHost -> "10"))
  val mongoRDD = sqlContext.fromMongoDB(mcInputBuilder.build())
  mongoRDD.registerTempTable(tableName)`

I've replaced the sampling ratio as well, but if you need to scan all your data for inferring the schema, please, restore it to 1.0. Take into account that you must upgrade to 0.11.0.

**Comment by Velmuruganms12** [comment]
thansk for the information will try upgrading to 0.11.0 and try this configuration
